crates/cheatcodes/src/strategy.rs

```diff
@@ -0,0 +1,327 @@
+        ccx: &mut CheatsCtxt,
+        account: Address,
+        new_nonce: u64,
+    ) -> Result {
+        let account = journaled_account(ccx.ecx, account)?;
+        // nonce must increment only
+        let current = account.info.nonce;
+        ensure!(
+            new_nonce >= current,
+            "new nonce ({new_nonce}) must be strictly equal to or higher than the \
+             account's current nonce ({current})"
+        );
+        account.info.nonce = new_nonce;
+        Ok(Default::default())
+    }
+
+    /// Cheatcode: setNonceUnsafe.
+    fn cheatcode_set_nonce_unsafe(
+        &mut self,
+        ccx: &mut CheatsCtxt,
+        account: Address,
+        new_nonce: u64,
+    ) -> Result {
+        let account = journaled_account(ccx.ecx, account)?;
+        account.info.nonce = new_nonce;
+        Ok(Default::default())
+    }
+
+    /// Mocks a call to return with a value.
+    fn mock_call(
+        &mut self,
+        ccx: &mut CheatsCtxt,
+        callee: Address,
+        data: &Bytes,
+        return_data: &Bytes,
+    ) -> Result {
+        let _ = make_acc_non_empty(&callee, ccx.ecx)?;
+        mock_call(ccx.state, &callee, data, None, return_data, InstructionResult::Return);
+        Ok(Default::default())
+    }
+
+    /// Mocks a call to revert with a value.
+    fn mock_call_revert(
+        &mut self,
+        ccx: &mut CheatsCtxt,
+        callee: Address,
+        data: &Bytes,
+        revert_data: &Bytes,
+    ) -> Result {
+        let _ = make_acc_non_empty(&callee, ccx.ecx)?;
+        mock_call(ccx.state, &callee, data, None, revert_data, InstructionResult::Revert);
+        Ok(Default::default())
+    }
+
+    /// Retrieve artifact code.
+    fn get_artifact_code(&self, state: &Cheatcodes, path: &str, deployed: bool) -> Result {
+        Ok(crate::fs::get_artifact_code(state, path, deployed)?.abi_encode())
+    }
+
+    /// Record broadcastable transaction during CREATE.
+    fn record_broadcastable_create_transactions(
+        &mut self,
+        config: Arc<CheatsConfig>,
+        input: &dyn CommonCreateInput,
+        ecx_inner: InnerEcx,
+        broadcast: &Broadcast,
+        broadcastable_transactions: &mut BroadcastableTransactions,
+    );
+
+    /// Record broadcastable transaction during CALL.
+    fn record_broadcastable_call_transactions(
+        &mut self,
+        config: Arc<CheatsConfig>,
+        input: &CallInputs,
+        ecx_inner: InnerEcx,
+        broadcast: &Broadcast,
+        broadcastable_transactions: &mut BroadcastableTransactions,
+        active_delegation: &mut Option<SignedAuthorization>,
+    );
+
+    fn post_initialize_interp(&mut self, _interpreter: &mut Interpreter, _ecx: Ecx) {}
+
+    /// Used to override opcode behaviors. Returns true if handled.
+    fn pre_step_end(&mut self, _interpreter: &mut Interpreter, _ecx: Ecx) -> bool {
+        false
+    }
+}
+
+/// We define this in our fork
+pub trait CheatcodeInspectorStrategyExt: CheatcodeInspectorStrategy {
```

Comment:
If that is the case, unless there's a hidden plan to make all zksync part of upstream with this flow, maybe we can just use `CheatcodeInspectorStrategy` as an owned type and we aggregate the fields we want on our fork? This way we don't go through the `Arc<Mutex>` stuff. A non fork integrated thing could define the strategy via a macro instead.

Reply:
The motivation is that the normal `CheatcodeInspectorStrategy`/`BackendStrategy`/`ExecutorStrategy` should contain the generic chain-agnostic methods in upstream that any chain can implement to allow foundry support in their own implementation. In an ideal scenario this would involve:

1. Fork foundry to foundry-foobar
2. Implement strategies in a separate `foundry-strategy-foobar` crate.
3. Modify `utils::get_executor_strategy(...)` to construct the either EVM or Foobar strategy based on some arbitrary config.

However, it would probably be the case that certain functionality cannot be abstracted to upstream (like our own zksync specific methods) and the implementors can declare those under `trait CheatcodeInspectorStrategyExt: CheatcodeInspectorStrategy` to and use the Ext trait everywhere. 
The other alternate being that we simply add the `zksync_` methods directly to `CheatcodeInspectorStrategy`. Maybe that's simpler as it doesn't involve replacing the strategy everywhere. But obviously this is a tradeoff, each fork can decide on their own.

Reply:
Makes sense. YMMV but given that the will be a fork, what would be ROI in defining traits? Wouldn't there always be a single trait implementation of each strategy? The flow minght be simpler if `Strategy` objects are just owned types, then forks just implement wrappers around them and use those wrappers.
So say for us and `CheatcodeInspectorStrategy` we would define `zksync::CheatcodeInspectorStrategy` that has the evm one in one of its fields and we replace it wherever the latter is called in upstream. Don't see the drawback of this, however it would simplify the flow by removing dynamic dispatch.

Reply:
> Wouldn't there always be a single trait implementation of each strategy

Realistically there'd always be at least two implementations in each foundry fork. If we abstract the strategies right, then the functionality can merely be accomplished with implementing the standard methods. Probably no one would define traits and we can go by adding them directly to the base strategy as well. 

For using concrete types, while _it can_ be done, it doesn't convey the intent that these concrete types are meant to be overridable. Plus one would need to replace the concrete type with the wrapped type across the code base. That's the downsides I can think of.

Reply:
> Realistically there'd always be at least two implementations in each foundry fork 
It would be two for `Strategy` (upstream + fork) and one for `StrategyExt` (fork, upstream does not implement this really) . And for the former you would actually instantiate only the fork one and call upstream because you wrap it right?

> That's the downsides I can think of.
Yeah those are fair, I just think they are cheaper than having to deal with `Arc<Mutex>` overall as things can sort of stay the same upstream. But well it is a matter of perspective. Just wanted to bring the owned type idea to the table.